{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "import json\n",
    "import yaml\n",
    "\n",
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "import pyspark.sql.types as t\n",
    "import pyspark.sql.functions as f\n",
    "\n",
    "import glow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/nasif12/home_if12/hoelzlwi/Projects/ALS/workflow/notebooks/../Snakefile'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snakefile_path = os.getcwd() + \"/../Snakefile\"\n",
    "snakefile_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# del snakemake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    snakemake\n",
    "except NameError:\n",
    "    from snakemk_util import load_rule_args\n",
    "\n",
    "    snakemake = load_rule_args(\n",
    "        snakefile = snakefile_path,\n",
    "        rule_name = 'annotate_variants_with_vep',\n",
    "        default_wildcards={\n",
    "        }\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"resources\": {\n",
      "    \"_cores\": 1,\n",
      "    \"_nodes\": 1\n",
      "  },\n",
      "  \"input\": {\n",
      "    \"vcf\": \"/s/raw/als/kaggle/end-als/genomics-data/AnswerALS_subset_annovar.hg38_anno_and_geno.no_intergenic.vcf.gz\",\n",
      "    \"gtf\": \"../data/common/Homo_sapiens.GRCh38.84.gtf\",\n",
      "    \"fasta\": \"/s/genomes/human/hg38/GRCh38.primary_assembly.genome.fa\"\n",
      "  },\n",
      "  \"params\": {\n",
      "    \"human_genome_version\": \"hg38\",\n",
      "    \"assembly\": \"GRCh38\",\n",
      "    \"vep_cache\": \"/opt/modules/i12g/conda-ensembl-vep/99/cachedir\",\n",
      "    \"loftee_data_dir\": \"/s/raw/loftee/GRCh38\",\n",
      "    \"cadd_dir\": \"/s/raw/cadd/v1.6/GRCh38\",\n",
      "    \"gnomad_vcf\": \"/s/raw/gnomad/2.1.1/hg38/gnomad.genomes.r2.1.1.sites.vcf.gz\",\n",
      "    \"vep_annotate_cmd\": \"/data/nasif12/home_if12/hoelzlwi/Projects/ALS/workflow/scripts/vep_99_skipheader.sh\"\n",
      "  },\n",
      "  \"output\": {\n",
      "    \"vep\": \"/s/project/kaggle-als/vep_annotations/ensembl_bychr.parquet\"\n",
      "  },\n",
      "  \"wildcards\": {}\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(json.dumps(snakemake.__dict__, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "memory: 184320000k\n",
      "number of cores: 110\n"
     ]
    }
   ],
   "source": [
    "MEM = os.popen(\"ulimit -m\").read()\n",
    "if MEM.startswith(\"unlimited\"):\n",
    "    print(\"Memory not constrained, using all available memory...\")\n",
    "    import psutil\n",
    "    MEM = psutil.virtual_memory().available / 1024\n",
    "MEM = int(MEM)\n",
    "\n",
    "N_CPU = int(os.popen(\"nproc\").read())\n",
    "\n",
    "print(\"memory: %dk\" % MEM)\n",
    "print(\"number of cores: %d\" % N_CPU)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MEM = int(MEM * 0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "                <p><b>SparkSession - in-memory</b></p>\n",
       "                \n",
       "        <div>\n",
       "            <p><b>SparkContext</b></p>\n",
       "\n",
       "            <p><a href=\"http://ouga05.cmm.in.tum.de:4040\">Spark UI</a></p>\n",
       "\n",
       "            <dl>\n",
       "              <dt>Version</dt>\n",
       "                <dd><code>v3.1.1</code></dd>\n",
       "              <dt>Master</dt>\n",
       "                <dd><code>local[110,4]</code></dd>\n",
       "              <dt>AppName</dt>\n",
       "                <dd><code>desmi_inject_gnomad</code></dd>\n",
       "            </dl>\n",
       "        </div>\n",
       "        \n",
       "            </div>\n",
       "        "
      ],
      "text/plain": [
       "<pyspark.sql.session.SparkSession at 0x7fce695d8040>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.environ['PYSPARK_SUBMIT_ARGS'] = \" \".join([\n",
    "    '--driver-memory %dk' % MEM,\n",
    "    'pyspark-shell'\n",
    "])\n",
    "os.environ['PYSPARK_SUBMIT_ARGS']\n",
    "\n",
    "MAX_FAILURES=4\n",
    "\n",
    "spark = (\n",
    "    SparkSession.builder\n",
    "    .appName('desmi_inject_gnomad')\n",
    "    .config(\"spark.jars.packages\", \",\".join([\n",
    "        \"io.projectglow:glow-spark3_2.12:1.0.0\",\n",
    "    ]))\n",
    "    .config(\"spark.local.dir\", os.environ.get(\"TMP\"))\n",
    "    .config(\"spark.master\", f\"local[{N_CPU},{MAX_FAILURES}]\")\n",
    "    .config(\"spark.sql.shuffle.partitions\", \"2001\")\n",
    "    .config(\"spark.sql.execution.arrow.enabled\", \"true\")\n",
    "    .config(\"spark.driver.maxResultSize\", \"48G\")\n",
    "    .config(\"spark.task.maxFailures\", MAX_FAILURES)\n",
    "    .getOrCreate()\n",
    ")\n",
    "glow.register(spark)\n",
    "spark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/s/raw/als/kaggle/end-als/genomics-data/AnswerALS_subset_annovar.hg38_anno_and_geno.no_intergenic.vcf.gz'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "INPUT_VCF  = snakemake.input[\"vcf\"]\n",
    "INPUT_VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/s/project/kaggle-als/vep_annotations/ensembl_bychr.parquet'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OUTPUT_PQ = snakemake.output[\"vep\"]\n",
    "OUTPUT_PQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/nasif12/home_if12/hoelzlwi/Projects/ALS/workflow/scripts/vep_99_skipheader.sh --output_file STDOUT --format vcf --cache --offline --dir=/opt/modules/i12g/conda-ensembl-vep/99/cachedir --force_overwrite --no_stats --tab --merged --assembly GRCh38 --fasta /s/genomes/human/hg38/GRCh38.primary_assembly.genome.fa --species homo_sapiens --everything --allele_number --total_length --numbers --symbol --hgvs --ccds --uniprot --af --af_1kg --af_esp --af_gnomad --max_af --pubmed --canonical --biotype --sift b --polyphen b --appris --domains --protein --regulatory --tsl --plugin Condel --plugin MaxEntScan,/opt/modules/i12g/conda-ensembl-vep/99/cachedir/Plugins/99_GRCh38/loftee/maxEntScan --plugin Blosum62 --plugin miRNA --plugin CADD,/s/raw/cadd/v1.6/GRCh38/whole_genome_SNVs.tsv.gz,/s/raw/cadd/v1.6/GRCh38/gnomad.genomes.r3.0.indel.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "FASTA=snakemake.input[\"fasta\"]\n",
    "GTF=snakemake.input[\"gtf\"]\n",
    "\n",
    "HUMAN_GENOME_VERSION=snakemake.params[\"human_genome_version\"]\n",
    "ASSEMBLY=snakemake.params[\"assembly\"]\n",
    "\n",
    "VEP_CACHE=snakemake.params[\"vep_cache\"]\n",
    "LOFTEE_DATA_DIR=snakemake.params[\"loftee_data_dir\"]\n",
    "CADD_DIR=snakemake.params[\"cadd_dir\"]\n",
    "GNOMAD_VCF=snakemake.params[\"gnomad_vcf\"]\n",
    "VEP_CMD = snakemake.params[\"vep_annotate_cmd\"]\n",
    "\n",
    "VEP_PLUGIN_DIR=f\"{VEP_CACHE}/Plugins/99_{ASSEMBLY}\"\n",
    "\n",
    "LOFTEE_PATH=f\"{VEP_PLUGIN_DIR}/loftee\"\n",
    "MAXENTSCAN_DATA_DIR=f\"{LOFTEE_PATH}/maxEntScan\"\n",
    "\n",
    "CADD_WGS_SNV=f\"{CADD_DIR}/whole_genome_SNVs.tsv.gz\"\n",
    "CADD_INDEL={\n",
    "    \"GRCh37\": f\"{CADD_DIR}/InDels.tsv.gz\",\n",
    "    \"GRCh38\": f\"{CADD_DIR}/gnomad.genomes.r3.0.indel.tsv.gz\",\n",
    "}[ASSEMBLY]\n",
    "\n",
    "vep_cmd=\" \".join([\n",
    "#     \"/opt/modules/i12g/anaconda/3-2019.10/bin/conda run --prefix /opt/modules/i12g/conda-ensembl-vep/99\",\n",
    "#     \"vep\",\n",
    "    VEP_CMD,\n",
    "    \"--output_file STDOUT\",\n",
    "    \"--format vcf\",\n",
    "    f\"--cache --offline --dir={VEP_CACHE}\",\n",
    "    \"--force_overwrite\",\n",
    "    \"--no_stats\",\n",
    "    \"--tab\",\n",
    "    \"--merged\",\n",
    "    f\"--assembly {ASSEMBLY}\",\n",
    "    f\"--fasta {FASTA}\",\n",
    "    \"--species homo_sapiens\",\n",
    "    \"--everything\",\n",
    "    \"--allele_number\",\n",
    "    \"--total_length\",\n",
    "    \"--numbers\",\n",
    "    \"--symbol\",\n",
    "    \"--hgvs\",\n",
    "    \"--ccds\",\n",
    "    \"--uniprot\",\n",
    "    \"--af\",\n",
    "    \"--af_1kg\",\n",
    "    \"--af_esp\",\n",
    "    \"--af_gnomad\",\n",
    "    \"--max_af\",\n",
    "    \"--pubmed\",\n",
    "    \"--canonical\",\n",
    "    \"--biotype\",\n",
    "    \"--sift b\",\n",
    "    \"--polyphen b\",\n",
    "    \"--appris\",\n",
    "    \"--domains\",\n",
    "    \"--protein\",\n",
    "    \"--regulatory\",\n",
    "    \"--tsl\",\n",
    "# Does not really work with hg38\n",
    "#     \"--plugin \" + \",\".join([\n",
    "#         \"LoF\",\n",
    "#         f\"loftee_path:{LOFTEE_PATH}\",\n",
    "#         f\"gerp_bigwig:{LOFTEE_DATA_DIR}/gerp_conservation_scores.homo_sapiens.{ASSEMBLY}.bw\",\n",
    "#         f\"human_ancestor_fa:{LOFTEE_DATA_DIR}/human_ancestor.fa.gz\",\n",
    "#         f\"conservation_file:{LOFTEE_DATA_DIR}/loftee.sql\",\n",
    "#     ]),\n",
    "#     f\"--plugin LoF,human_ancestor_fa:{LOFTEE_DATA_DIR}/human_ancestor.fa.gz,conservation_file:{LOFTEE_DATA_DIR}/loftee.sql,loftee_path:{LOFTEE_PATH}\",\n",
    "    \"--plugin Condel\",\n",
    "    f\"--plugin MaxEntScan,{MAXENTSCAN_DATA_DIR}\",\n",
    "    \"--plugin Blosum62\",\n",
    "    \"--plugin miRNA\",\n",
    "    f\"--plugin CADD,{CADD_WGS_SNV},{CADD_INDEL}\",\n",
    "])\n",
    "#    --gtf $GTF\n",
    "#    --plugin LoFtool\n",
    "#    --plugin dbscSNV\n",
    "#    --custom ${GNOMAD_VCF},Gnomad_2.1.1,vcf,overlap\n",
    "#    --plugin AncestralAllele\n",
    "\n",
    "print(vep_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (\n",
    "    spark\n",
    "    .read\n",
    "    .option(\"flattenInfoFields\", False)\n",
    "    .format('vcf')\n",
    "    .load(INPUT_VCF)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = glow.transform(\"split_multiallelics\", df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- contigName: string (nullable = true)\n",
      " |-- start: long (nullable = true)\n",
      " |-- end: long (nullable = true)\n",
      " |-- names: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- referenceAllele: string (nullable = true)\n",
      " |-- alternateAlleles: array (nullable = false)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- qual: double (nullable = true)\n",
      " |-- filters: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- splitFromMultiAllelic: boolean (nullable = false)\n",
      " |-- attributes: map (nullable = true)\n",
      " |    |-- key: string\n",
      " |    |-- value: string (valueContainsNull = true)\n",
      " |-- INFO_OLD_MULTIALLELIC: string (nullable = true)\n",
      " |-- genotypes: array (nullable = true)\n",
      " |    |-- element: struct (containsNull = false)\n",
      " |    |    |-- sampleId: string (nullable = true)\n",
      " |    |    |-- conditionalQuality: integer (nullable = true)\n",
      " |    |    |-- MQ0: integer (nullable = true)\n",
      " |    |    |-- alleleDepths: array (nullable = true)\n",
      " |    |    |    |-- element: integer (containsNull = true)\n",
      " |    |    |-- PID: string (nullable = true)\n",
      " |    |    |-- phased: boolean (nullable = true)\n",
      " |    |    |-- calls: array (nullable = true)\n",
      " |    |    |    |-- element: integer (containsNull = true)\n",
      " |    |    |-- PGT: string (nullable = true)\n",
      " |    |    |-- phredLikelihoods: array (nullable = true)\n",
      " |    |    |    |-- element: integer (containsNull = true)\n",
      " |    |    |-- depth: integer (nullable = true)\n",
      " |    |    |-- AB: double (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.withColumn(\"names\", f.array([f.concat(\n",
    "    f.col('contigName'),\n",
    "    f.lit(\":\"),\n",
    "    f.col('start') + 1,\n",
    "    f.lit(\":\"),\n",
    "    f.col('referenceAllele'),\n",
    "    f.lit(\">\"),\n",
    "    f.col('alternateAlleles')[0]\n",
    ")]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/florian3/lib/python3.8/site-packages/pyspark/sql/pandas/conversion.py:87: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.pyspark.enabled' is set to true; however, failed by the reason below:\n",
      "  Unsupported type in conversion to Arrow: ArrayType(StructType(List(StructField(sampleId,StringType,true),StructField(conditionalQuality,IntegerType,true),StructField(MQ0,IntegerType,true),StructField(alleleDepths,ArrayType(IntegerType,true),true),StructField(PID,StringType,true),StructField(phased,BooleanType,true),StructField(calls,ArrayType(IntegerType,true),true),StructField(PGT,StringType,true),StructField(phredLikelihoods,ArrayType(IntegerType,true),true),StructField(depth,IntegerType,true),StructField(AB,DoubleType,true))),false)\n",
      "Attempting non-optimization as 'spark.sql.execution.arrow.pyspark.fallback.enabled' is set to true.\n",
      "  warnings.warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contigName</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>names</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAlleles</th>\n",
       "      <th>qual</th>\n",
       "      <th>filters</th>\n",
       "      <th>splitFromMultiAllelic</th>\n",
       "      <th>attributes</th>\n",
       "      <th>INFO_OLD_MULTIALLELIC</th>\n",
       "      <th>genotypes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16377</td>\n",
       "      <td>16378</td>\n",
       "      <td>[chr1:16378:T&gt;C]</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>53076.99</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [39, 19], None...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16486</td>\n",
       "      <td>16487</td>\n",
       "      <td>[chr1:16487:T&gt;C]</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>15370.21</td>\n",
       "      <td>[VQSRTrancheSNP99.00to99.90]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [61, 0], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16494</td>\n",
       "      <td>16495</td>\n",
       "      <td>[chr1:16495:G&gt;C]</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>63161.88</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [67, 20], None...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16533</td>\n",
       "      <td>16534</td>\n",
       "      <td>[chr1:16534:C&gt;T]</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>19334.20</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 94, None, [55, 11], None...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16566</td>\n",
       "      <td>16567</td>\n",
       "      <td>[chr1:16567:G&gt;C]</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>270.35</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [56, 0], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16570</td>\n",
       "      <td>16571</td>\n",
       "      <td>[chr1:16571:G&gt;A]</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>3495.12</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 92, None, [54, 3], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16575</td>\n",
       "      <td>16576</td>\n",
       "      <td>[chr1:16576:A&gt;C]</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>52.18</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [41, 0], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16630</td>\n",
       "      <td>16631</td>\n",
       "      <td>[chr1:16631:T&gt;C]</td>\n",
       "      <td>T</td>\n",
       "      <td>[C]</td>\n",
       "      <td>132.14</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 99, None, [34, 0], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16681</td>\n",
       "      <td>16682</td>\n",
       "      <td>[chr1:16682:G&gt;A]</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1147.17</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 37, None, [28, 5], None,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16687</td>\n",
       "      <td>16688</td>\n",
       "      <td>[chr1:16688:G&gt;A]</td>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1188.62</td>\n",
       "      <td>[VQSRTrancheSNP99.90to100.00]</td>\n",
       "      <td>False</td>\n",
       "      <td>{'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...</td>\n",
       "      <td>None</td>\n",
       "      <td>[(CGND-HDA-00187-b38, 30, None, [33, 3], None,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  contigName  start    end             names referenceAllele alternateAlleles  \\\n",
       "0       chr1  16377  16378  [chr1:16378:T>C]               T              [C]   \n",
       "1       chr1  16486  16487  [chr1:16487:T>C]               T              [C]   \n",
       "2       chr1  16494  16495  [chr1:16495:G>C]               G              [C]   \n",
       "3       chr1  16533  16534  [chr1:16534:C>T]               C              [T]   \n",
       "4       chr1  16566  16567  [chr1:16567:G>C]               G              [C]   \n",
       "5       chr1  16570  16571  [chr1:16571:G>A]               G              [A]   \n",
       "6       chr1  16575  16576  [chr1:16576:A>C]               A              [C]   \n",
       "7       chr1  16630  16631  [chr1:16631:T>C]               T              [C]   \n",
       "8       chr1  16681  16682  [chr1:16682:G>A]               G              [A]   \n",
       "9       chr1  16687  16688  [chr1:16688:G>A]               G              [A]   \n",
       "\n",
       "       qual                        filters  splitFromMultiAllelic  \\\n",
       "0  53076.99  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "1  15370.21   [VQSRTrancheSNP99.00to99.90]                  False   \n",
       "2  63161.88  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "3  19334.20  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "4    270.35  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "5   3495.12  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "6     52.18  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "7    132.14  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "8   1147.17  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "9   1188.62  [VQSRTrancheSNP99.90to100.00]                  False   \n",
       "\n",
       "                                          attributes INFO_OLD_MULTIALLELIC  \\\n",
       "0  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "1  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "2  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "3  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "4  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "5  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "6  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "7  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "8  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "9  {'BS2': '.', 'BS1': '.', 'BS4': '.', 'ExAC_non...                  None   \n",
       "\n",
       "                                           genotypes  \n",
       "0  [(CGND-HDA-00187-b38, 99, None, [39, 19], None...  \n",
       "1  [(CGND-HDA-00187-b38, 99, None, [61, 0], None,...  \n",
       "2  [(CGND-HDA-00187-b38, 99, None, [67, 20], None...  \n",
       "3  [(CGND-HDA-00187-b38, 94, None, [55, 11], None...  \n",
       "4  [(CGND-HDA-00187-b38, 99, None, [56, 0], None,...  \n",
       "5  [(CGND-HDA-00187-b38, 92, None, [54, 3], None,...  \n",
       "6  [(CGND-HDA-00187-b38, 99, None, [41, 0], None,...  \n",
       "7  [(CGND-HDA-00187-b38, 99, None, [34, 0], None,...  \n",
       "8  [(CGND-HDA-00187-b38, 37, None, [28, 5], None,...  \n",
       "9  [(CGND-HDA-00187-b38, 30, None, [33, 3], None,...  "
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.limit(10).toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import shlex\n",
    "\n",
    "input_df = df.select([\n",
    "    f.col('contigName'),\n",
    "    f.col('start'),\n",
    "    f.col('end'),\n",
    "    f.col('names'),\n",
    "    f.col('referenceAllele'),\n",
    "    f.col('alternateAlleles'),\n",
    "])\n",
    "\n",
    "vep_transformed_df = glow.transform(\n",
    "    \"pipe\",\n",
    "    input_df,\n",
    "#     cmd=json.dumps(shlex.split(\"cat | grep -v '^##'\")),\n",
    "    cmd=json.dumps(shlex.split(vep_cmd)),\n",
    "    inputFormatter='vcf',\n",
    "    inVcfHeader='infer',\n",
    "    outputFormatter='csv',\n",
    "#     outQuote=\"##\",\n",
    "    outHeader=True,\n",
    "    outDelimiter=\"\\t\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- #Uploaded_variation: string (nullable = true)\n",
      " |-- Location: string (nullable = true)\n",
      " |-- Allele: string (nullable = true)\n",
      " |-- Gene: string (nullable = true)\n",
      " |-- Feature: string (nullable = true)\n",
      " |-- Feature_type: string (nullable = true)\n",
      " |-- Consequence: string (nullable = true)\n",
      " |-- cDNA_position: string (nullable = true)\n",
      " |-- CDS_position: string (nullable = true)\n",
      " |-- Protein_position: string (nullable = true)\n",
      " |-- Amino_acids: string (nullable = true)\n",
      " |-- Codons: string (nullable = true)\n",
      " |-- Existing_variation: string (nullable = true)\n",
      " |-- ALLELE_NUM: string (nullable = true)\n",
      " |-- IMPACT: string (nullable = true)\n",
      " |-- DISTANCE: string (nullable = true)\n",
      " |-- STRAND: string (nullable = true)\n",
      " |-- FLAGS: string (nullable = true)\n",
      " |-- VARIANT_CLASS: string (nullable = true)\n",
      " |-- SYMBOL: string (nullable = true)\n",
      " |-- SYMBOL_SOURCE: string (nullable = true)\n",
      " |-- HGNC_ID: string (nullable = true)\n",
      " |-- BIOTYPE: string (nullable = true)\n",
      " |-- CANONICAL: string (nullable = true)\n",
      " |-- MANE: string (nullable = true)\n",
      " |-- TSL: string (nullable = true)\n",
      " |-- APPRIS: string (nullable = true)\n",
      " |-- CCDS: string (nullable = true)\n",
      " |-- ENSP: string (nullable = true)\n",
      " |-- SWISSPROT: string (nullable = true)\n",
      " |-- TREMBL: string (nullable = true)\n",
      " |-- UNIPARC: string (nullable = true)\n",
      " |-- REFSEQ_MATCH: string (nullable = true)\n",
      " |-- SOURCE: string (nullable = true)\n",
      " |-- GIVEN_REF: string (nullable = true)\n",
      " |-- USED_REF: string (nullable = true)\n",
      " |-- BAM_EDIT: string (nullable = true)\n",
      " |-- GENE_PHENO: string (nullable = true)\n",
      " |-- SIFT: string (nullable = true)\n",
      " |-- PolyPhen: string (nullable = true)\n",
      " |-- EXON: string (nullable = true)\n",
      " |-- INTRON: string (nullable = true)\n",
      " |-- DOMAINS: string (nullable = true)\n",
      " |-- miRNA43: string (nullable = true)\n",
      " |-- HGVSc: string (nullable = true)\n",
      " |-- HGVSp: string (nullable = true)\n",
      " |-- HGVS_OFFSET: string (nullable = true)\n",
      " |-- AF: string (nullable = true)\n",
      " |-- AFR_AF: string (nullable = true)\n",
      " |-- AMR_AF: string (nullable = true)\n",
      " |-- EAS_AF: string (nullable = true)\n",
      " |-- EUR_AF: string (nullable = true)\n",
      " |-- SAS_AF: string (nullable = true)\n",
      " |-- AA_AF: string (nullable = true)\n",
      " |-- EA_AF: string (nullable = true)\n",
      " |-- gnomAD_AF: string (nullable = true)\n",
      " |-- gnomAD_AFR_AF: string (nullable = true)\n",
      " |-- gnomAD_AMR_AF: string (nullable = true)\n",
      " |-- gnomAD_ASJ_AF: string (nullable = true)\n",
      " |-- gnomAD_EAS_AF: string (nullable = true)\n",
      " |-- gnomAD_FIN_AF: string (nullable = true)\n",
      " |-- gnomAD_NFE_AF: string (nullable = true)\n",
      " |-- gnomAD_OTH_AF: string (nullable = true)\n",
      " |-- gnomAD_SAS_AF: string (nullable = true)\n",
      " |-- MAX_AF: string (nullable = true)\n",
      " |-- MAX_AF_POPS: string (nullable = true)\n",
      " |-- CLIN_SIG: string (nullable = true)\n",
      " |-- SOMATIC: string (nullable = true)\n",
      " |-- PHENO: string (nullable = true)\n",
      " |-- PUBMED: string (nullable = true)\n",
      " |-- MOTIF_NAME: string (nullable = true)\n",
      " |-- MOTIF_POS: string (nullable = true)\n",
      " |-- HIGH_INF_POS: string (nullable = true)\n",
      " |-- MOTIF_SCORE_CHANGE: string (nullable = true)\n",
      " |-- LoF: string (nullable = true)\n",
      " |-- LoF_filter: string (nullable = true)\n",
      " |-- LoF_flags: string (nullable = true)\n",
      " |-- LoF_info: string (nullable = true)\n",
      " |-- Condel: string (nullable = true)\n",
      " |-- MaxEntScan_alt: string (nullable = true)\n",
      " |-- MaxEntScan_diff: string (nullable = true)\n",
      " |-- MaxEntScan_ref: string (nullable = true)\n",
      " |-- BLOSUM62: string (nullable = true)\n",
      " |-- miRNA83: string (nullable = true)\n",
      " |-- CADD_PHRED: string (nullable = true)\n",
      " |-- CADD_RAW: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "vep_transformed_df.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- chrom: string (nullable = true)\n",
      " |-- pos: string (nullable = true)\n",
      " |-- ref: string (nullable = true)\n",
      " |-- alt: string (nullable = true)\n",
      " |-- Gene: string (nullable = true)\n",
      " |-- Feature: string (nullable = true)\n",
      " |-- Feature_type: string (nullable = true)\n",
      " |-- Consequence: string (nullable = true)\n",
      " |-- cDNA_position: string (nullable = true)\n",
      " |-- CDS_position: string (nullable = true)\n",
      " |-- Protein_position: string (nullable = true)\n",
      " |-- Amino_acids: string (nullable = true)\n",
      " |-- Codons: string (nullable = true)\n",
      " |-- Existing_variation: string (nullable = true)\n",
      " |-- ALLELE_NUM: string (nullable = true)\n",
      " |-- IMPACT: string (nullable = true)\n",
      " |-- DISTANCE: string (nullable = true)\n",
      " |-- STRAND: string (nullable = true)\n",
      " |-- FLAGS: string (nullable = true)\n",
      " |-- VARIANT_CLASS: string (nullable = true)\n",
      " |-- SYMBOL: string (nullable = true)\n",
      " |-- SYMBOL_SOURCE: string (nullable = true)\n",
      " |-- HGNC_ID: string (nullable = true)\n",
      " |-- BIOTYPE: string (nullable = true)\n",
      " |-- CANONICAL: boolean (nullable = false)\n",
      " |-- MANE: string (nullable = true)\n",
      " |-- TSL: string (nullable = true)\n",
      " |-- APPRIS: string (nullable = true)\n",
      " |-- CCDS: string (nullable = true)\n",
      " |-- ENSP: string (nullable = true)\n",
      " |-- SWISSPROT: string (nullable = true)\n",
      " |-- TREMBL: string (nullable = true)\n",
      " |-- UNIPARC: string (nullable = true)\n",
      " |-- REFSEQ_MATCH: string (nullable = true)\n",
      " |-- SOURCE: string (nullable = true)\n",
      " |-- GIVEN_REF: string (nullable = true)\n",
      " |-- USED_REF: string (nullable = true)\n",
      " |-- BAM_EDIT: string (nullable = true)\n",
      " |-- GENE_PHENO: string (nullable = true)\n",
      " |-- EXON: string (nullable = true)\n",
      " |-- INTRON: string (nullable = true)\n",
      " |-- DOMAINS: string (nullable = true)\n",
      " |-- miRNA43: string (nullable = true)\n",
      " |-- HGVSc: string (nullable = true)\n",
      " |-- HGVSp: string (nullable = true)\n",
      " |-- HGVS_OFFSET: string (nullable = true)\n",
      " |-- AF: string (nullable = true)\n",
      " |-- AFR_AF: string (nullable = true)\n",
      " |-- AMR_AF: string (nullable = true)\n",
      " |-- EAS_AF: string (nullable = true)\n",
      " |-- EUR_AF: string (nullable = true)\n",
      " |-- SAS_AF: string (nullable = true)\n",
      " |-- AA_AF: string (nullable = true)\n",
      " |-- EA_AF: string (nullable = true)\n",
      " |-- gnomAD_AF: string (nullable = true)\n",
      " |-- gnomAD_AFR_AF: string (nullable = true)\n",
      " |-- gnomAD_AMR_AF: string (nullable = true)\n",
      " |-- gnomAD_ASJ_AF: string (nullable = true)\n",
      " |-- gnomAD_EAS_AF: string (nullable = true)\n",
      " |-- gnomAD_FIN_AF: string (nullable = true)\n",
      " |-- gnomAD_NFE_AF: string (nullable = true)\n",
      " |-- gnomAD_OTH_AF: string (nullable = true)\n",
      " |-- gnomAD_SAS_AF: string (nullable = true)\n",
      " |-- MAX_AF: string (nullable = true)\n",
      " |-- MAX_AF_POPS: string (nullable = true)\n",
      " |-- CLIN_SIG: string (nullable = true)\n",
      " |-- SOMATIC: string (nullable = true)\n",
      " |-- PHENO: string (nullable = true)\n",
      " |-- PUBMED: string (nullable = true)\n",
      " |-- MOTIF_NAME: string (nullable = true)\n",
      " |-- MOTIF_POS: string (nullable = true)\n",
      " |-- HIGH_INF_POS: string (nullable = true)\n",
      " |-- MOTIF_SCORE_CHANGE: string (nullable = true)\n",
      " |-- LoF: string (nullable = true)\n",
      " |-- LoF_filter: string (nullable = true)\n",
      " |-- LoF_flags: string (nullable = true)\n",
      " |-- LoF_info: string (nullable = true)\n",
      " |-- MaxEntScan_alt: string (nullable = true)\n",
      " |-- MaxEntScan_diff: string (nullable = true)\n",
      " |-- MaxEntScan_ref: string (nullable = true)\n",
      " |-- BLOSUM62: string (nullable = true)\n",
      " |-- miRNA83: string (nullable = true)\n",
      " |-- CADD_PHRED: string (nullable = true)\n",
      " |-- CADD_RAW: string (nullable = true)\n",
      " |-- condel_prediction: string (nullable = true)\n",
      " |-- condel_score: string (nullable = true)\n",
      " |-- sift_prediction: string (nullable = true)\n",
      " |-- sift_score: string (nullable = true)\n",
      " |-- polyphen_prediction: string (nullable = true)\n",
      " |-- polyphen_score: string (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "parsed_df = (\n",
    "    vep_transformed_df\n",
    "    .withColumn(\"#Uploaded_variation\", f.split(f.col(\"#Uploaded_variation\"), '[:>]'))\n",
    "    .select([\n",
    "        f.col(\"#Uploaded_variation\")[0].alias(\"chrom\"),\n",
    "        f.col(\"#Uploaded_variation\")[1].alias(\"pos\"),\n",
    "        f.col(\"#Uploaded_variation\")[2].alias(\"ref\"),\n",
    "        f.col(\"#Uploaded_variation\")[3].alias(\"alt\"),\n",
    "        \"*\",\n",
    "    ])\n",
    "    .withColumn(\"ref\", f.when(f.col(\"ref\") == '-', \"\").otherwise(f.col(\"ref\")))\n",
    "    .withColumn(\"alt\", f.when(f.col(\"alt\") == '-', \"\").otherwise(f.col(\"alt\")))\n",
    "    .withColumn(\"CANONICAL\", f.when(f.col(\"CANONICAL\") == 'YES', True).otherwise(False))\n",
    "    .withColumn(\"condel_prediction\", f.regexp_replace(f.col(\"Condel\"), \"\\\\(.*\", \"\"))\n",
    "    .withColumn(\"condel_score\", f.regexp_replace(f.col(\"Condel\"), \"(.*\\\\()|\\\\)\", \"\"))\n",
    "    .withColumn(\"sift_prediction\", f.regexp_replace(f.col(\"SIFT\"), \"\\\\(.*\", \"\"))\n",
    "    .withColumn(\"sift_score\", f.regexp_replace(f.col(\"SIFT\"), \"(.*\\\\()|\\\\)\", \"\"))\n",
    "    .withColumn(\"polyphen_prediction\", f.regexp_replace(f.col(\"PolyPhen\"), \"\\\\(.*\", \"\"))\n",
    "    .withColumn(\"polyphen_score\", f.regexp_replace(f.col(\"PolyPhen\"), \"(.*\\\\()|\\\\)\", \"\"))\n",
    "    .drop(\n",
    "        \"#Uploaded_variation\", \n",
    "        \"Location\", \n",
    "        \"Allele\",\n",
    "        \"Condel\",\n",
    "        \"SIFT\",\n",
    "        \"PolyPhen\",\n",
    "    )\n",
    ")\n",
    "parsed_df.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "# x = parsed_df.filter(f.col(\"Condel\") != '-').limit(10).toPandas()\n",
    "# x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [],
   "source": [
    "dtypes = {\n",
    "    \"consequence\": t.ArrayType(t.StringType()),\n",
    "    \"Existing_variation\": t.ArrayType(t.StringType()),\n",
    "    \"ALLELE_NUM\": t.IntegerType(),\n",
    "    \"DISTANCE\": t.IntegerType(),\n",
    "    \"STRAND\": t.ShortType(),\n",
    "    \"FLAGS\": t.ArrayType(t.StringType()),\n",
    "    \"HGNC_ID\": t.IntegerType(),\n",
    "#     \"CANONICAL\": t.BooleanType(), # needs manual check if column equals \"CANONICAL\"\n",
    "    \"TREMBL\": t.ArrayType(t.StringType()),\n",
    "    \"REFSEQ_MATCH\": t.ArrayType(t.StringType()),\n",
    "    \"GENE_PHENO\": t.BooleanType(),\n",
    "    \"sift_score\": t.FloatType(),\n",
    "    \"polyphen_score\": t.FloatType(),\n",
    "    \"EXON\": t.ArrayType(t.IntegerType()),\n",
    "    \"INTRON\": t.ArrayType(t.IntegerType()),\n",
    "    \"HGVS_OFFSET\": t.IntegerType(),\n",
    "    \"AF\": t.ArrayType(t.FloatType()),\n",
    "    \"AFR_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"AMR_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"EAS_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"EUR_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"SAS_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"AA_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"EA_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_AFR_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_AMR_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_ASJ_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_EAS_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_FIN_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_NFE_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_OTH_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"gnomAD_SAS_AF\": t.ArrayType(t.FloatType()),\n",
    "    \"MAX_AF\": t.FloatType(),\n",
    "    \"MAX_AF_POPS\": t.ArrayType(t.StringType()),\n",
    "    \"PUBMED\": t.ArrayType(t.StringType()),\n",
    "    \"MOTIF_POS\": t.IntegerType(),\n",
    "    \"MOTIF_SCORE_CHANGE\": t.FloatType(),\n",
    "    \"Condel\": t.StringType(),\n",
    "    \"condel_score\": t.FloatType(),\n",
    "    \"condel_prediction\": t.StringType(),\n",
    "    \"BLOSUM62\": t.IntegerType(),\n",
    "    \"LoF\": t.StringType(),\n",
    "    \"LoF_filter\": t.StringType(),\n",
    "    \"LoF_flags\": t.StringType(),\n",
    "    \"MaxEntScan_ref\": t.FloatType(),\n",
    "    \"MaxEntScan_alt\": t.FloatType(),\n",
    "    \"MaxEntScan_diff\": t.FloatType(),\n",
    "    \"CADD_PHRED\": t.FloatType(),\n",
    "    \"CADD_RAW\": t.FloatType(),\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [],
   "source": [
    "needsMinVal = {\n",
    "    \"AF\",\n",
    "    \"AFR_AF\",\n",
    "    \"AMR_AF\",\n",
    "    \"EAS_AF\",\n",
    "    \"EUR_AF\",\n",
    "    \"SAS_AF\",\n",
    "    \"AA_AF\",\n",
    "    \"EA_AF\",\n",
    "    \"gnomAD_AF\",\n",
    "    \"gnomAD_AFR_AF\",\n",
    "    \"gnomAD_AMR_AF\",\n",
    "    \"gnomAD_ASJ_AF\",\n",
    "    \"gnomAD_EAS_AF\",\n",
    "    \"gnomAD_FIN_AF\",\n",
    "    \"gnomAD_NFE_AF\",\n",
    "    \"gnomAD_OTH_AF\",\n",
    "    \"gnomAD_SAS_AF\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_col(name):\n",
    "    col = f.col(name)\n",
    "    col = f.when(col == '-', None).otherwise(col)\n",
    "    if name in dtypes:\n",
    "        dtype = dtypes[name]\n",
    "        if isinstance(dtype, t.ArrayType):\n",
    "            col = f.split(col, \",\")\n",
    "        col = col.cast(dtypes[name])\n",
    "    if name in needsMinVal:\n",
    "        col = f.array_min(col)\n",
    "    \n",
    "    return col.alias(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [],
   "source": [
    "parsed_df = parsed_df.select(*[\n",
    "    parse_col(x) for x in parsed_df.columns\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- chrom: string (nullable = true)\n",
      " |-- pos: string (nullable = true)\n",
      " |-- ref: string (nullable = true)\n",
      " |-- alt: string (nullable = true)\n",
      " |-- Gene: string (nullable = true)\n",
      " |-- Feature: string (nullable = true)\n",
      " |-- Feature_type: string (nullable = true)\n",
      " |-- Consequence: string (nullable = true)\n",
      " |-- cDNA_position: string (nullable = true)\n",
      " |-- CDS_position: string (nullable = true)\n",
      " |-- Protein_position: string (nullable = true)\n",
      " |-- Amino_acids: string (nullable = true)\n",
      " |-- Codons: string (nullable = true)\n",
      " |-- Existing_variation: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- ALLELE_NUM: integer (nullable = true)\n",
      " |-- IMPACT: string (nullable = true)\n",
      " |-- DISTANCE: integer (nullable = true)\n",
      " |-- STRAND: short (nullable = true)\n",
      " |-- FLAGS: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- VARIANT_CLASS: string (nullable = true)\n",
      " |-- SYMBOL: string (nullable = true)\n",
      " |-- SYMBOL_SOURCE: string (nullable = true)\n",
      " |-- HGNC_ID: integer (nullable = true)\n",
      " |-- BIOTYPE: string (nullable = true)\n",
      " |-- CANONICAL: boolean (nullable = true)\n",
      " |-- MANE: string (nullable = true)\n",
      " |-- TSL: string (nullable = true)\n",
      " |-- APPRIS: string (nullable = true)\n",
      " |-- CCDS: string (nullable = true)\n",
      " |-- ENSP: string (nullable = true)\n",
      " |-- SWISSPROT: string (nullable = true)\n",
      " |-- TREMBL: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- UNIPARC: string (nullable = true)\n",
      " |-- REFSEQ_MATCH: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- SOURCE: string (nullable = true)\n",
      " |-- GIVEN_REF: string (nullable = true)\n",
      " |-- USED_REF: string (nullable = true)\n",
      " |-- BAM_EDIT: string (nullable = true)\n",
      " |-- GENE_PHENO: boolean (nullable = true)\n",
      " |-- EXON: array (nullable = true)\n",
      " |    |-- element: integer (containsNull = true)\n",
      " |-- INTRON: array (nullable = true)\n",
      " |    |-- element: integer (containsNull = true)\n",
      " |-- DOMAINS: string (nullable = true)\n",
      " |-- miRNA43: string (nullable = true)\n",
      " |-- HGVSc: string (nullable = true)\n",
      " |-- HGVSp: string (nullable = true)\n",
      " |-- HGVS_OFFSET: integer (nullable = true)\n",
      " |-- AF: float (nullable = true)\n",
      " |-- AFR_AF: float (nullable = true)\n",
      " |-- AMR_AF: float (nullable = true)\n",
      " |-- EAS_AF: float (nullable = true)\n",
      " |-- EUR_AF: float (nullable = true)\n",
      " |-- SAS_AF: float (nullable = true)\n",
      " |-- AA_AF: float (nullable = true)\n",
      " |-- EA_AF: float (nullable = true)\n",
      " |-- gnomAD_AF: float (nullable = true)\n",
      " |-- gnomAD_AFR_AF: float (nullable = true)\n",
      " |-- gnomAD_AMR_AF: float (nullable = true)\n",
      " |-- gnomAD_ASJ_AF: float (nullable = true)\n",
      " |-- gnomAD_EAS_AF: float (nullable = true)\n",
      " |-- gnomAD_FIN_AF: float (nullable = true)\n",
      " |-- gnomAD_NFE_AF: float (nullable = true)\n",
      " |-- gnomAD_OTH_AF: float (nullable = true)\n",
      " |-- gnomAD_SAS_AF: float (nullable = true)\n",
      " |-- MAX_AF: float (nullable = true)\n",
      " |-- MAX_AF_POPS: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- CLIN_SIG: string (nullable = true)\n",
      " |-- SOMATIC: string (nullable = true)\n",
      " |-- PHENO: string (nullable = true)\n",
      " |-- PUBMED: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- MOTIF_NAME: string (nullable = true)\n",
      " |-- MOTIF_POS: integer (nullable = true)\n",
      " |-- HIGH_INF_POS: string (nullable = true)\n",
      " |-- MOTIF_SCORE_CHANGE: float (nullable = true)\n",
      " |-- LoF: string (nullable = true)\n",
      " |-- LoF_filter: string (nullable = true)\n",
      " |-- LoF_flags: string (nullable = true)\n",
      " |-- LoF_info: string (nullable = true)\n",
      " |-- MaxEntScan_alt: float (nullable = true)\n",
      " |-- MaxEntScan_diff: float (nullable = true)\n",
      " |-- MaxEntScan_ref: float (nullable = true)\n",
      " |-- BLOSUM62: integer (nullable = true)\n",
      " |-- miRNA83: string (nullable = true)\n",
      " |-- CADD_PHRED: float (nullable = true)\n",
      " |-- CADD_RAW: float (nullable = true)\n",
      " |-- condel_prediction: string (nullable = true)\n",
      " |-- condel_score: float (nullable = true)\n",
      " |-- sift_prediction: string (nullable = true)\n",
      " |-- sift_score: float (nullable = true)\n",
      " |-- polyphen_prediction: string (nullable = true)\n",
      " |-- polyphen_score: float (nullable = true)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "parsed_df.printSchema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>...</th>\n",
       "      <th>BLOSUM62</th>\n",
       "      <th>miRNA83</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>condel_prediction</th>\n",
       "      <th>condel_score</th>\n",
       "      <th>sift_prediction</th>\n",
       "      <th>sift_score</th>\n",
       "      <th>polyphen_prediction</th>\n",
       "      <th>polyphen_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000335137</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>80/1054</td>\n",
       "      <td>44/918</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>16.910</td>\n",
       "      <td>1.683178</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.319</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.07</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>167/2618</td>\n",
       "      <td>107/981</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>16.910</td>\n",
       "      <td>1.683178</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.322</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.07</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>79501</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>44/918</td>\n",
       "      <td>44/918</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>16.910</td>\n",
       "      <td>1.683178</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.319</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.07</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69428</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000335137</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>374/1054</td>\n",
       "      <td>338/918</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>24.500</td>\n",
       "      <td>3.441204</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.836</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.01</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69428</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>461/2618</td>\n",
       "      <td>401/981</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>24.500</td>\n",
       "      <td>3.441204</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.883</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.02</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69428</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>79501</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>338/918</td>\n",
       "      <td>338/918</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>None</td>\n",
       "      <td>24.500</td>\n",
       "      <td>3.441204</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.836</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.01</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69511</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000335137</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>457/1054</td>\n",
       "      <td>421/918</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>2.209</td>\n",
       "      <td>0.112916</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.82</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69511</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000641515</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>544/2618</td>\n",
       "      <td>484/981</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>2.209</td>\n",
       "      <td>0.112916</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.85</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69511</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>79501</td>\n",
       "      <td>NM_001005484.1</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>421/918</td>\n",
       "      <td>421/918</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>2.209</td>\n",
       "      <td>0.112916</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.001</td>\n",
       "      <td>tolerated</td>\n",
       "      <td>0.82</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>69761</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>ENST00000335137</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>707/1054</td>\n",
       "      <td>671/918</td>\n",
       "      <td>...</td>\n",
       "      <td>-3</td>\n",
       "      <td>None</td>\n",
       "      <td>14.750</td>\n",
       "      <td>1.351540</td>\n",
       "      <td>neutral</td>\n",
       "      <td>0.406</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.02</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows  90 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    pos ref alt             Gene          Feature Feature_type  \\\n",
       "0  chr1  69134   A   G  ENSG00000186092  ENST00000335137   Transcript   \n",
       "1  chr1  69134   A   G  ENSG00000186092  ENST00000641515   Transcript   \n",
       "2  chr1  69134   A   G            79501   NM_001005484.1   Transcript   \n",
       "3  chr1  69428   T   G  ENSG00000186092  ENST00000335137   Transcript   \n",
       "4  chr1  69428   T   G  ENSG00000186092  ENST00000641515   Transcript   \n",
       "5  chr1  69428   T   G            79501   NM_001005484.1   Transcript   \n",
       "6  chr1  69511   A   G  ENSG00000186092  ENST00000335137   Transcript   \n",
       "7  chr1  69511   A   G  ENSG00000186092  ENST00000641515   Transcript   \n",
       "8  chr1  69511   A   G            79501   NM_001005484.1   Transcript   \n",
       "9  chr1  69761   A   T  ENSG00000186092  ENST00000335137   Transcript   \n",
       "\n",
       "        Consequence cDNA_position CDS_position  ... BLOSUM62 miRNA83  \\\n",
       "0  missense_variant       80/1054       44/918  ...       -2    None   \n",
       "1  missense_variant      167/2618      107/981  ...       -2    None   \n",
       "2  missense_variant        44/918       44/918  ...       -2    None   \n",
       "3  missense_variant      374/1054      338/918  ...       -2    None   \n",
       "4  missense_variant      461/2618      401/981  ...       -2    None   \n",
       "5  missense_variant       338/918      338/918  ...       -2    None   \n",
       "6  missense_variant      457/1054      421/918  ...        0    None   \n",
       "7  missense_variant      544/2618      484/981  ...        0    None   \n",
       "8  missense_variant       421/918      421/918  ...        0    None   \n",
       "9  missense_variant      707/1054      671/918  ...       -3    None   \n",
       "\n",
       "  CADD_PHRED  CADD_RAW  condel_prediction condel_score  sift_prediction  \\\n",
       "0     16.910  1.683178            neutral        0.319        tolerated   \n",
       "1     16.910  1.683178            neutral        0.322        tolerated   \n",
       "2     16.910  1.683178            neutral        0.319        tolerated   \n",
       "3     24.500  3.441204        deleterious        0.836      deleterious   \n",
       "4     24.500  3.441204        deleterious        0.883      deleterious   \n",
       "5     24.500  3.441204        deleterious        0.836      deleterious   \n",
       "6      2.209  0.112916            neutral        0.001        tolerated   \n",
       "7      2.209  0.112916            neutral        0.001        tolerated   \n",
       "8      2.209  0.112916            neutral        0.001        tolerated   \n",
       "9     14.750  1.351540            neutral        0.406      deleterious   \n",
       "\n",
       "   sift_score polyphen_prediction polyphen_score  \n",
       "0        0.07              benign          0.000  \n",
       "1        0.07              benign          0.045  \n",
       "2        0.07              benign          0.000  \n",
       "3        0.01   probably_damaging          0.984  \n",
       "4        0.02   probably_damaging          1.000  \n",
       "5        0.01   probably_damaging          0.984  \n",
       "6        0.82              benign          0.000  \n",
       "7        0.85              benign          0.000  \n",
       "8        0.82              benign          0.000  \n",
       "9        0.02              benign          0.229  \n",
       "\n",
       "[10 rows x 90 columns]"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parsed_df.filter(~ f.isnull(\"condel_score\")).limit(10).toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [],
   "source": [
    "# (\n",
    "#     parsed_df\n",
    "#     .filter(f.col(\"Gene\").startswith(\"ENSG\"))\n",
    "#     .sortWithinPartitions([\"chrom\", \"pos\"])\n",
    "#     .write\n",
    "#     .mode(\"overwrite\")\n",
    "#     .partitionBy(\"Gene\")\n",
    "#     .parquet(OUTPUT_PATH + \"/ensembl_genes.parquet\")\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    parsed_df\n",
    "    .filter(f.col(\"Gene\").startswith(\"ENSG\"))\n",
    "    .sortWithinPartitions([\"chrom\", \"Gene\", \"pos\"]) # hierarchical sorting\n",
    "    .write\n",
    "    .mode(\"overwrite\")\n",
    "    .partitionBy(\"chrom\")\n",
    "    .parquet(OUTPUT_PQ)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_json": true,
   "formats": "ipynb,py"
  },
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-florian3]",
   "language": "python",
   "name": "conda-env-anaconda-florian3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
